### PR TITLE
[autoscaler] Add AWS Autoscaler CloudWatch Alarm support

### PR DESCRIPTION
**Pull request by Zyiqin-Miranda**
## Why are these changes needed?

These changes add a set of improvements to enable automatic creation and update of CloudWatch alarms when provisioning AWS Autoscaling clusters. Successful implementation of these improvements will allow AWS Autoscaler users to:
   1. Setup alarms against Ray CloudWatch metrics to get notified about increased load, service outage.
   2. Update their CloudWatch alarm JSON configuration files during Ray up execution time.
  
Notes:

This PR is a follow-up PR for #20266, which adds CloudWatch alarm support.

## Related issue number

"Closes #9644 https://github.com/ray-project/ray/issues/8967"

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py [hunk 1]
```diff
@@ -420,6 +447,23 @@ def _replace_dashboard_config_variables(self) -> Dict[str, Any]:
             widgets.append(item_out)
         return widgets
 
+    def _replace_alarm_config_variables(self) -> Dict[str, Any]:
```
Comment (pdames) [inline python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py:450]: Should the return type here be `List[Dict[str, Any]]`?

File: python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py [hunk 2]
```diff
@@ -90,6 +93,8 @@ def _update_cloudwatch_config(self, is_head_node: bool,
                         self._restart_cloudwatch_agent()
                     elif config_type == "dashboard":
                         self._put_cloudwatch_dashboard()
+                    elif config_type == "alarm":
```
Comment (pdames) [inline python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py:96]: Minor: Maybe we could add an enum class to standardize repeated references to agent/dashboard/alarm strings?

```python
class ConfigType(str, Enum):
    AGENT = "agent"
    DASHBOARD = "dashboard"
    ALARM = "alarm"
```

File: python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py [hunk 3]
```diff
@@ -420,6 +447,23 @@ def _replace_dashboard_config_variables(self) -> Dict[str, Any]:
             widgets.append(item_out)
         return widgets
 
+    def _replace_alarm_config_variables(self) -> Dict[str, Any]:
+        """
+        replace known variable occurrences in cloudwatch alarm config file
```
Comment (pdames) [inline python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py:452]: Can we enumerate what the known variables are in this docstring (instance ID, cluster name, and region)?

**Review by pdames — COMMENTED** [review]
LGTM overall - just a few minor suggestions!

@wuisawesome and @DmitriGekhtman - can you also take a look?

File: python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py (hunk 1 above)
Comment (Zyiqin-Miranda) [reply python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py]: Yes, it should be List[Dict[str, Any]], corrected.

File: python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py (hunk 3 above)
Comment (Zyiqin-Miranda) [reply python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py]: Sure, added variables description in _replace_cwa_config_variables(), _replace_dashboard_config_variables() and _replace_alarm_config_variables() functions.

File: python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py (hunk 2 above)
Comment (Zyiqin-Miranda) [reply python/ray/autoscaler/_private/aws/cloudwatch/cloudwatch_helper.py:96]: Thanks for the enum idea, I'll do it in the next PR that consolidates all three config types and removes redundancy.

**Review by pdames — APPROVED** [review]

**Review by DmitriGekhtman — APPROVED** [review]
@wuisawesome could you take a final look?

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by Zyiqin-Miranda** [comment]
Resolved merge conflicts from switching code formatter from YAPF to Black.

**Review by wuisawesome — APPROVED** [review]

**Comment by wuisawesome** [comment]
(whoops not sure why i added that label) 

rerunning the tests now since they were failling due to an infra failure, the will merge if they still look good

**Comment by wuisawesome** [comment]
Lint failure was broken in master, java failure is unrelated/flakey. Merging.

**Pull request merged**